drivers/mtd_spi_nor/mtd_spi_nor.c

```diff
@@ -0,0 +1,489 @@
+    uint32_t page_addr_mask = dev->page_addr_mask;
+    if ((addr & page_addr_mask) != ((addr + size - 1) & page_addr_mask)) {
+        /* Reads across page boundaries must be split */
+        size = mtd->page_size - (addr & ~(page_addr_mask));
+    }
+    if (size == 0) {
+        return 0;
+    }
+    be_uint32_t addr_be = byteorder_htonl(addr);
+    mtd_spi_cmd_addr_read(dev, dev->opcode->read, addr_be, dest, size);
+
+    return size;
+}
+
+static int mtd_spi_nor_write(mtd_dev_t *mtd, const void *src, uint32_t addr, uint32_t size)
+{
+    DEBUG("mtd_spi_nor_write: %p, %p, 0x%" PRIx32 ", 0x%" PRIx32 "\n",
+        (void *)mtd, src, addr, size);
+    if (size == 0) {
+        return 0;
+    }
+    mtd_spi_nor_t *dev = (mtd_spi_nor_t *)mtd;
+    if (size > mtd->page_size) {
+        DEBUG("mtd_spi_nor_write: ERR: page program >1 page (%" PRIu32 ")!\n", mtd->page_size);
+        return -EOVERFLOW;
+    }
+    if (dev->page_addr_mask &&
+        ((addr & dev->page_addr_mask) != ((addr + size - 1) & dev->page_addr_mask))) {
+        DEBUG("mtd_spi_nor_write: ERR: page program spans page boundary!\n");
+        return -EOVERFLOW;
+    }
+    be_uint32_t addr_be = byteorder_htonl(addr);
+
+    /* write enable */
+    mtd_spi_cmd(dev, dev->opcode->wren);
+
+    /* Page program */
+    mtd_spi_cmd_addr_write(dev, dev->opcode->page_program, addr_be, src, size);
+
+    /* waiting for the command to complete before returning */
+    wait_for_write_complete(dev);
+    return size;
+}
+
+static int mtd_spi_nor_erase(mtd_dev_t *mtd, uint32_t addr, uint32_t size)
+{
+    DEBUG("mtd_spi_nor_erase: %p, 0x%" PRIx32 ", 0x%" PRIx32 "\n",
+        (void *)mtd, addr, size);
+    mtd_spi_nor_t *dev = (mtd_spi_nor_t *)mtd;
+    uint32_t total_size = mtd->page_size * mtd->pages_per_sector * mtd->sector_count;
+
+    if (dev->sec_addr_mask &&
+        ((addr & ~dev->sec_addr_mask) != 0)) {
+        /* This is not a requirement in hardware, but it helps in catching
+         * software bugs (the erase-all-your-files kind) */
+        DEBUG("addr = %" PRIx32 " ~dev->erase_addr_mask = %" PRIx32 "", addr, ~dev->sec_addr_mask);
+        DEBUG("mtd_spi_nor_erase: ERR: erase addr not aligned on %" PRIu32 " byte boundary.\n",
+            sector_size);
+        return -EINVAL;
+    }
+    be_uint32_t addr_be = byteorder_htonl(addr);
+
+    /* write enable */
+    mtd_spi_cmd(dev, dev->opcode->wren);
+
+    if (size == total_size) {
+        mtd_spi_cmd_addr_write(dev, dev->opcode->chip_erase, addr_be, NULL, 0);
+    }
+    else if ((dev->flag & SPI_NOR_F_SECT_4K) && size == 4096) {
+        /* 4 KiO sectors can be erased with sector erase command */
+        mtd_spi_cmd_addr_write(dev, dev->opcode->sector_erase, addr_be, NULL, 0);
+    }
+    else if ((dev->flag & SPI_NOR_F_SECT_32K) && size == 32768) {
+        /* 4 KiO sectors can be erased with sector erase command */
+        mtd_spi_cmd_addr_write(dev, dev->opcode->block_erase_32k, addr_be, NULL, 0);
+    }
+    else {
```

Comment:
`size` is not checked here. This causes the unit test for erase of invalid size to fail.